lib/needlessDisables.js

```diff
@@ -1,95 +1,92 @@
 
 /**
  * @param {import('stylelint').StylelintResult[]} results
- * @returns {StylelintDisableOptionsReport}
  */
 module.exports = function (results) {
-	/** @type {StylelintDisableOptionsReport} */
-	const report = [];
-
 	results.forEach((result) => {
 		// File with `CssSyntaxError` have not `_postcssResult`
 		if (!result._postcssResult) {
 			return;
 		}
 
-		/** @type {{ranges: DisableReportRange[], source: string}} */
-		const unused = { source: result.source || '', ranges: [] };
-
-		/** @type {{[ruleName: string]: Array<RangeType>}} */
-		const rangeData = _.cloneDeep(result._postcssResult.stylelint.disabledRanges);
+		/** @type {{[ruleName: string]: Array<DisabledRange>}} */
+		const rangeData = result._postcssResult.stylelint.disabledRanges;
 
 		if (!rangeData) {
 			return;
 		}
 
 		const disabledWarnings = result._postcssResult.stylelint.disabledWarnings || [];
 
+		// A map from `stylelint-disable` comments to the set of rules that
+		// are usefully disabled by each comment. We track this
+		// comment-by-comment rather than range-by-range because ranges that
+		// disable *all* rules are duplicated for each rule they apply to in
+		// practice.
+		/** @type {Map<PostcssComment, Set<string>>}} */
+		const usefulDisables = new Map();
+
 		disabledWarnings.forEach((warning) => {
 			const rule = warning.rule;
-
 			const ruleRanges = rangeData[rule];
 
 			if (ruleRanges) {
-				// Back to front so we get the *last* range that applies to the warning
-				for (const range of ruleRanges.reverse()) {
+				for (const range of ruleRanges) {
 					if (isWarningInRange(warning, range)) {
-						range.used = true;
-
-						return;
+						putIfAbsent(usefulDisables, range.comment, () => new Set()).add(rule);
 					}
 				}
 			}
 
-			for (const range of rangeData.all.reverse()) {
+			for (const range of rangeData.all) {
 				if (isWarningInRange(warning, range)) {
-					range.used = true;
-
-					return;
+					putIfAbsent(usefulDisables, range.comment, () => new Set()).add(rule);
 				}
 			}
 		});
 
-		Object.keys(rangeData).forEach((rule) => {
-			rangeData[rule].forEach((range) => {
-				// Is an equivalent range already marked as unused?
-				const alreadyMarkedUnused = unused.ranges.find((unusedRange) => {
-					return unusedRange.start === range.start && unusedRange.end === range.end;
-				});
+		// Get rid of the duplicated ranges for each `all` rule. We only care
+		// if the entire `all` rule is useful as a whole or not.
+		for (const range of rangeData.all) {
+			for (const [rule, ranges] of Object.entries(rangeData)) {
+				if (rule === 'all') continue;
 
-				// If this range is unused and no equivalent is marked,
-				// mark this range as unused
-				if (!range.used && !alreadyMarkedUnused) {
-					unused.ranges.push({
-						rule,
-						start: range.start,
-						end: range.end,
-						unusedRule: rule,
-					});
-				}
+				_.remove(ranges, (otherRange) => range.comment === otherRange.comment);
```

Comment:
not sure that mutating `result._postcssResult.stylelint.disabledRanges` is fine

Reply:
Good point, I should be cloning that field.